docs/cudf/source/user_guide/copy-on-write.md

```diff
@@ -0,0 +1,175 @@
+(copy-on-write-user-doc)=
+
+# Copy-on-write
+
+Copy-on-write is a memory management strategy that allows multiple cuDF objects containing the same data to refer to the same memory address as long as neither of them modify the underlying data.
+With this approach, any operation that generates an unmodified view of an object (such as copies, slices, or methods like `DataFrame.head`) returns a new object that points to the same memory as the original.
+However, when either the existing or new object is _modified_, a copy of the data is made prior to the modification, ensuring that the changes do not propagate between the two objects.
+This behavior is best understood by looking at the examples below.
+
+The default behaviour in cuDF is for copy-on-write to be disabled, so to use it, one must explicitly
+opt in by setting a cuDF option. It is recommended to set the copy-on-write at the beginning of the
+script execution, because when this setting is changed in middle of a script execution there will
+be un-intended behavior where the objects created when copy-on-write is enabled will still have the
+copy-on-write behavior whereas the objects created when copy-on-write is disabled will have
+different behavior.
+
+## Enabling copy-on-write
+
+1. Use `cudf.set_option`:
+
+    ```python
+    >>> import cudf
+    >>> cudf.set_option("copy_on_write", True)
+    ```
+
+2. Set the environment variable ``CUDF_COPY_ON_WRITE`` to ``1`` prior to the
+launch of the Python interpreter:
+
+    ```bash
+    export CUDF_COPY_ON_WRITE="1" python -c "import cudf"
+    ```
+
+## Disabling copy-on-write
+
+
+Copy-on-write can be disabled by setting the ``copy_on_write`` option to ``False``:
+
+```python
+>>> cudf.set_option("copy_on_write", False)
+```
+
+## Making copies
+
+There are no additional changes required in the code to make use of copy-on-write.
+
+```python
+>>> series = cudf.Series([1, 2, 3, 4])
+```
+
+Performing a shallow copy will create a new Series object pointing to the
+same underlying device memory:
+
+```python
+>>> copied_series = series.copy(deep=False)
+>>> series
+0    1
+1    2
+2    3
+3    4
+dtype: int64
+>>> copied_series
+0    1
+1    2
+2    3
+3    4
+dtype: int64
+```
+
+When a write operation is performed on either ``series`` or
+``copied_series``, a true physical copy of the data is created:
+
+```python
+>>> series[0:2] = 10
+>>> series
+0    10
+1    10
+2     3
+3     4
+dtype: int64
+>>> copied_series
+0    1
+1    2
+2    3
+3    4
+dtype: int64
+```
+
+
+## Notes
+
+When copy-on-write is enabled, there is no concept of views. Modifying any view created inside cuDF will always trigger a copy and will not modify the original object.
```

Comment:
Do you suggest modifying the above to the following?:

```
When copy-on-write is enabled, modifying any view created inside cuDF will always trigger a copy and will not modify the original object.
```

Reply:
Yeah I had this question too during an earlier round of review, but I wasn't sure how to improve the phrasing here. Technically it is correct: `s2 = s.iloc[:]` creates a view, so future modifications of `s2` are modifications of a view, whereas `s.iloc[:] = 1` does not create any views but modifies `s` directly. I don't know if that's worth spelling out, or if it is a distinction we should expect users to understand.

Reply:
I guess this is because `s.iloc[:] = 1` calls `s.iloc.__setitem__(:, 1)`, whereas
```
s2 = s.iloc[:]
```
Calls `s.iloc.__getitem__(:)` which creates a new object (distinct from `s.iloc`) and now `s2.iloc[:] = 1` doesn't call `s.iloc.__setitem__`.

As noted with all of the copy-on-write discussions in pandas (and here) and all the explanations in https://pandas.pydata.org/docs/user_guide/indexing.html#indexing-view-versus-copy, the indexing on dataframes is actually really hard to grok in general.

In this particular case it's additionally confusing because the same expression can mean different things depending on whether it is an lvalue or rvalue.

In fact, with copy-on-write on, slicing behaves _much more_ like slicing on builtin python objects like lists. So perhaps we should say that, or something along those lines. People might be more used to that:
```python
x = [1,2, 3]
y = x[:] # copy!
y[0] = 5
assert x[0] == 1
```

So how about:

> When copy-on-write is enabled, there is no longer a concept of views when slicing or indexing. In this sense indexing behaves as one would expect for builtin Python containers like lists, rather than indexing numpy arrays. Modifying a "view" created by cuDF will always trigger a copy and will not modify the original object.

Reply:
Correct, this is fundamentally about `_setitem__` vs `__getitem__`. The rephrasing looks good to me.